homework-2/scripts/DomRenderer.js

```diff
@@ -0,0 +1,296 @@
+/*
+3. Implement DOM renderer:
+    1. You must create el function, which has the following signature
+    function el(type: string, attrs: object, children: DomElement | DomElement[])
+    So basically it accepts the following:
+    a. The type of the dom element to be created ('div', 'span', etc)
+    b. An object with tag attributes (class, id, etc)
+    c. Child elements. This one can be either DomElement (SpanElement for instance) or an array of DomElement.
+    In case of an array, all elements in the array are siblings under same parent
+    d. The el function should return a DomElement instance
+    2. Create DomElement class, which must be a base class for all Elements
+    3. Each class extending DomElement should implement draw method
+    4. Each HTML tag must have a corresponding class extending DomElement. For instance a "div" tag should have
+    DivElement class:
+    class DivElement extends DomElement {
+      draw(children) {
+        // 1. Create div with
+        // const div = document.createElement("DIV");
+        // 2. Append children to div
+        // 3. Return div
+      }
+    }
+*/
+
+class DomElement {
+  constructor(attrs, children) {
+    this.attributes = attrs;
+    this.children = children;
+
+    if (new.target === DomElement) {
+      throw new Error("Can not instantiate base class.");
+    }
+  }
+
+  draw(tagName) {
+    const domElement = document.createElement(tagName);
+    for (let attr in this.attributes) {
+      domElement.setAttribute(attr, this.attributes[attr]);
+    }
+
+    if (this.children) {
+      if (typeof this.children === "string") {
+        domElement.appendChild(document.createTextNode(this.children));
+      } else if (Array.isArray(this.children)) {
+        this.children.forEach((child) => domElement.appendChild(child.draw()));
+      } else if (this.children instanceof DomElement) {
+        domElement.appendChild(this.children.draw());
+      }
+    }
```

Comment:
looks like we have a lot of conditions related to children here. let's see how we can simplify this. ideally we would like to do this:
```
this.children.forEach(child => {
    domElement.appendChild(child.draw());
})
```
to achieve it, firstly this.children should always be an array, instead of being "null or a single element or an array". 
see if you can do that in the constructor.

secondly, instead of checking if the child is a string and creating a text node, let's just call draw() and let the child's object decide how to handle it. this means all the children need to be DomElement instances with a draw method, not just strings or numbers. this will become helpful if in the future we have more children with different ways to draw them - you would create a new class without needing to modify this part of code.

Reply:
Please review the changes.

Reply:
test